test/integration/controlplane/graceful_shutdown_test.go

```diff
@@ -70,7 +70,14 @@ func TestGracefulShutdown(t *testing.T) {
 	resp.Body.Close()
 
 	t.Logf("shutting down server")
-	tearDownOnce.Do(server.TearDownFn)
+	// We tear it down in the background to ensure that
+	// pending requests should work fine.
```

Comment:
I don't fully understand this change, before we "guarantee" that new requests come once the apiserver was down, is not going to race this way?

Reply:
Sorry - can you clarify your comment?

Before the changes I'm making in this PR to make "server.TearDownFn" actually be synchronous, this was only triggerring the shutdown and the lines below were happening in the meantime.

With "server.TearDownFn" being synchronous, when that is finishing, the server is already down. So there is no chance that writing anything to body in line 96 will succeed.
[What wouldn't be happening without this adjustment to test is that the request is hanging, server is waiting to be drained, but finally the requests is timing out and the server is the shutting down.]

In order to make it work, we need to trigger the shutdown here, but actually let it work in the background.

Reply:
ic, the thing that synchronizes with the shutdown is when it start to fail "new requests", so you know it started to shutdown.
Then, it waits 500 ms, write "garbage" and checks that it returns something 
Then, you wait for the shutdown to exit

Is there any way to guarantee that the shutdown never finish before we can write?

Reply:
> Is there any way to guarantee that the shutdown never finish before we can write?

In general the request that we're writing "garbage" to is started before here (line 44). So it already reached apiserver.

Part of the apiserver shutdown sequence is that it won't close the http server, before all requests are actually drained. In our case, it includes that particular request.

The request will not be drained until it finishes, so given we don't finish it before writing to it, the only other way for it to finish is when it times out on the server side.

So our only risk is request timing out on the server side. But the default timeouts are 1m, and I think it's fine to leave it that way.

[FWIW - this is not a regression in this PR - the same failure mode existed before (if we didn't write anything to the body for 1m and it would timeout on the server side.]

Reply:
yeah, I just wanted to fully understand the test and the change, thanks for the explanation